### PR TITLE
Add support for text detetction for multi entity multi text in ner_v2 

### PR DESCRIPTION
**Pull request by ankur09011**
## JIRA Ticket Number

JIRA TICKET: ML-1892, ML-1893, ML-1896

## Description of change

* Add new text detector module `text_detection.py` inside `ner_v2/detectors/textuals` 
* Add new endpoint for text detection for multiple entities inside urls file i.e `v2/text/
* Add API  view in `ner_v2/api` for text detection.
* Add test module for text_detection module and  utility function module.
## Checklist (OPTIONAL):

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: ner_v2/detectors/textual/utils.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
+import json
+import six
+
+from chatbot_ner.config import ner_logger
```
Comment (haptik-deployment) [inline ner_v2/detectors/textual/utils.py:4]: F401 'chatbot_ner.config.ner_logger' imported but unused

File: ner_v2/detectors/textual/utils.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    Parse text request coming from POST call on `/v2/text/`
+    Args:
+        request: request object
+
+    Returns:
+        output data
+    """
+    request_data = json.loads(request.body)
+    queries = request_data.get("queries")
+    language_script = request_data.get('language_script', ENGLISH_LANG)
```
Comment (haptik-deployment) [inline ner_v2/detectors/textual/utils.py:62]: F841 local variable 'language_script' is assigned to but never used

File: ner_v2/detectors/textual/utils.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        request: request object
+
+    Returns:
+        output data
+    """
+    request_data = json.loads(request.body)
+    message = request_data.get("message")
+    bot_message = request_data.get("bot_message")
+    entities = request_data.get("entities", {})
+    language_script = request_data.get('language_script', ENGLISH_LANG)
```
Comment (haptik-deployment) [inline ner_v2/detectors/textual/utils.py:128]: F841 local variable 'language_script' is assigned to but never used

File: ner_v2/detectors/textual/utils.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
+import json
```
Comment (haptik-deployment) [inline ner_v2/detectors/textual/utils.py:1]: W1618 [34mimport missing `from __future__ import absolute_import`[0m

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

File: ner_v2/detectors/textual/utils.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+                                    structured_value=None, bot_message=bot_message)
+
+        data[0]["entities"].update(output[0])
+
+        # get detection for structured value text entities
+        if structured_value_entities:
+            for entity, value in structured_value_entities.items():
+                entity_dict = {entity: value}
+                sv = value.get("structured_value")
+                print(sv)
```
Comment (haptik-deployment) [inline ner_v2/detectors/textual/utils.py:105]: E1601 [1;31mprint statement used[0m

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

File: ner_v2/detectors/textual/utils.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    Returns:
+        TextDetection output for default fallback
+    """
+    output = {}
+    if not entities_dict:
+        return output
+
+    for entity, value in entities_dict.items():
+        fallback_value = value.get("fallback_value")
+
```
Comment (haptik-deployment) [inline ner_v2/detectors/textual/utils.py:146]: W293 blank line contains whitespace

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

File: ner_v2/detectors/textual/tests/test_utils.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+            {'entity_value': {'value': 'Mumbai', 'datastore_verified': True,
+                              'model_verified': False},
+             'detection': 'message', 'original_text': 'mumbai',
+             'language': 'en'}], 'restaurant': []},
+            'language': 'en'}]
+
+        output = parse_text_request(request)
+
+        assert_output = [{'entities':
+            {'entities': {'city': [
```
Comment (haptik-deployment) [inline ner_v2/detectors/textual/tests/test_utils.py:104]: E128 continuation line under-indented for visual indent

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

File: ner_v2/detectors/textual/tests/test_text_detection.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+                             'datastore_verified': True, 'model_verified': False},
+            'detection': 'message', 'original_text': 'mumbai', 'language': 'en'}],
+            'restaurant': [
+                {'entity_value': {'value': "Domino's Pizza", 'datastore_verified': True, 'model_verified': False},
+                 'detection': 'message', 'original_text': 'dominoes', 'language': 'en'}]}]
+
+        self.maxDiff = None
+        self.assertListEqual(result, assert_output)
+
+    @patch('ner_v2.detectors.textual.elastic_search.'
```
Comment (haptik-deployment) [inline ner_v2/detectors/textual/tests/test_text_detection.py:90]: F811 redefinition of unused 'test_text_detection_detect_single_message' from line 45

File: ner_v2/detectors/textual/tests/test_text_detection.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+                                     target_language_script=target_language_script)
+
+        result = text_detector.detect_bulk(messages=message)
+
+        assert_output = [{'city': [
+            {'entity_value': {'value': 'Mumbai', 'datastore_verified': True, 'model_verified': False},
+             'detection': 'message', 'original_text': 'mumbai', 'language': 'en'}], 'restaurant': [
+            {'entity_value': {'value': "Domino's Pizza", 'datastore_verified': True, 'model_verified': False},
+             'detection': 'message', 'original_text': 'dominoes', 'language': 'en'}]}, {'city': [
+            {'entity_value': {'value': 'New Delhi', 'datastore_verified': True, 'model_verified': False},
```
Comment (haptik-deployment) [inline ner_v2/detectors/textual/tests/test_text_detection.py:135]: E122 continuation line missing indentation or outdented

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

File: ner_v2/api.py [hunk 10]
```diff
@@ -557,7 +559,137 @@ def phone_number(request):
...
+
+        entities (dict): dictionary of entties to be detected, each entity dict will contain
+            following details:
+
+            entity_name (str): name of the entity. Also acts as elastic-search dictionary name
+                              if entity uses elastic-search lookup
+            structured_value (str): [Optional] Value obtained from any structured elements. Note if structured value is
+                                   detection is run on structured value instead of message
+                                   (For example, UI elements like form, payload, etc)
+            fallback_value (str): [Optional] If the detection logic fails to detect any value either from structured_value
```
Comment (haptik-deployment) [inline ner_v2/api.py:586]: E501 line too long (122 > 119 characters)

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

File: ner_v2/detectors/textual/text_detection.py [hunk 11]
```diff
@@ -0,0 +1,706 @@
...
+                    return text[start:end]
+        except (ValueError, IndexError):
+            ner_logger.exception('Error getting original text (%s, %s)' % (matched_tokens, text))
+
+        return u' '.join(matched_tokens)
+
+    def _get_text_detection_with_variants(self):
+        """
+            This function will normalise the message by breaking it into trigrams, bigrams and unigrams. The generated
+            ngrams will be used to create query to retrieve search results from datastore. These results will contain list
```
Comment (haptik-deployment) [inline ner_v2/detectors/textual/text_detection.py:266]: E501 line too long (122 > 119 characters)

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

File: ner_v2/detectors/textual/text_detection.py [hunk 12]
```diff
@@ -0,0 +1,711 @@
...
+
+                            In auto mode, if length of term is less than low it must match exactly,
+                              if it is between [low, high) one insert/delete/substitution is allowed,
+                              for anything higher than equal to high, two inserts/deletes/substitutions
+                               are allowed
+
+                    `_min_token_size_for_fuzziness (int)`: minimum number of letters a word must
+                                have to be considered for calculating edit distance with similar
+                                ngrams from the datastore
+
```
Comment (haptik-deployment) [inline ner_v2/detectors/textual/text_detection.py:62]: W293 blank line contains whitespace

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS FAILED... Please resolve issues

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by ankur09011** [comment]
- Add structured query to single call for ES

File: ner_v2/detectors/textual/text_detection.py [hunk 13]
```diff
@@ -0,0 +1,827 @@
...
+            list of lists: list of dict for each message with key as entity name
+                            containing the detected text entities and original message.
+        """
+
+        self._process_text(messages)
+
+        texts = [u' '.join(TOKENIZER.tokenize(processed_text)) for
+                 processed_text in self.__processed_texts]
+
+        entities_dict = self.entities_dict
```
Comment (haptik-deployment) [inline ner_v2/detectors/textual/text_detection.py:430]: F841 local variable 'entities_dict' is assigned to but never used

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS FAILED... Please resolve issues

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

File: ner_v2/detectors/textual/tests/test_elastic_search.py [hunk 14]
```diff
@@ -62,3 +62,24 @@ def test_elasticsearch_get_dynamic_fuzziness_threshold(self):
...
+
+    def test_add_query(self):
+        es = ElasticSearchDataStore()
+
+        entity_list_1 = ['city', 'restaurant']
+        text_1 = "I want to go to mumbai"
+
+        query_data = es.generate_query_data(entities=entity_list_1, texts=text_1)
+
+        print(query_data)
```
Comment (haptik-deployment) [inline ner_v2/detectors/textual/tests/test_elastic_search.py:74]: E1601 [1;31mprint statement used[0m

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

File: ner_v2/api.py [hunk 15]
```diff
@@ -557,7 +559,143 @@ def phone_number(request):
...
+        response = {"success": False, "error": "Get method is not allowed"}
+        return HttpResponse(json.dumps(response), status=501)
+
+    elif request.method == "POST":
+        ner_logger.debug("Fetching result")
+
+        try:
+            verify_text_request(request)
+            # if verify success parse request and get data
+            data = parse_text_request(request)
```
Comment (amansrivastava17) [inline ner_v2/api.py:678]: Should be seperate out parsing and get data logic ? 
According to function definition it should do parsing job.

File: ner_v2/detectors/textual/utils.py [hunk 16]
```diff
@@ -0,0 +1,283 @@
...
+        ner_logger.exception("Entities param is not in correct format")
+        raise TypeError("Entities should be dict of entity details")
+
+    if len(message) > 100:
+        ner_logger.exception("Maximum number of message can be 100 for "
+                             "bulk detection")
+        raise ValueError("Maximum number of message can be 100 for "
+                         "bulk detection")
+
+    if len(list(entities)) > 100:
```
Comment (amansrivastava17) [inline ner_v2/detectors/textual/utils.py:58]: haptik_api does not have any batching on number of entities... should we add batching there ? or chatbot ner should handle ?

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS FAILED... Please resolve issues

File: ner_v2/detectors/textual/tests/test_text_detection.py [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+                                        ('G.', 'G  Pulla Reddy Sweets')]),
+             'city': OrderedDict([('Delhi', 'New Delhi'), ('Wani', 'Wani'),
+                                  ('goa', 'goa')])}]
+
+        text_detector = TextDetector(entity_dict=entity_dict, source_language_script=language,
+                                     target_language_script=target_language_script)
+
+        result = text_detector.detect_bulk(messages=message)
+
+        print(result)
```
Comment (haptik-deployment) [inline ner_v2/detectors/textual/tests/test_text_detection.py:130]: E1601 [1;31mprint statement used[0m

File: ner_v2/detectors/textual/utils.py (hunk 16 above)
Comment (ankur09011) [reply ner_v2/detectors/textual/utils.py:58]: I think we should have to batch at both places. Need to test for bulk entities for optimal numbers for batches.

File: ner_v2/api.py (hunk 15 above)
Comment (ankur09011) [reply ner_v2/api.py]: Agree, `verify_text_request` is parsing and verifying the contents.
Changed the name of function and it's definition.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS FAILED... Please resolve issues

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Review by amansrivastava17 — APPROVED** [review]

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Comment by haptik-deployment** [comment]
:+1: No lint errors found.

**Comment by haptik-deployment** [comment]
UNIT TESTS HAVE PASSED... Good To Merge

**Review by amansrivastava17 — APPROVED** [review]

**Pull request merged**